pint/formatting.py

```diff
@@ -117,6 +117,135 @@ def _pretty_fmt_exponent(num):
     },
 }
 
+#: _FORMATTERS maps format names to callables doing the formatting
+_FORMATTERS: Dict[str, Callable] = {}
+
+
+def register_unit_format(name):
+    """register a function as a new format for units
+
+    The registered function must have a signature of:
+
+    .. code:: python
+
+        def new_format(unit, registry, **options):
+            pass
+
+    Parameters
+    ----------
+    name : str
+        The name of the new format (to be used in the format mini-language). A error is
+        raised if the new format would overwrite a existing format.
+
+    Examples
+    --------
+    >>> @pint.register_unit_format("custom")
+    ... def format_custom(unit, registry, **options):
+    ...     result = "<formatted unit>"  # do the formatting
+    ...     return result
+    ...
+    >>> ureg = pint.UnitRegistry()
+    >>> u = ureg.m / ureg.s ** 2
+    >>> f"{u:custom}"
+    '<formatted unit>'
+    """
+
+    def wrapper(func):
+        if name in _FORMATTERS:
+            raise ValueError(f"format {name:!r} already exists")  # or warn instead
```

Comment:
should this be a warning or an error?

Reply:
I think it should produce an error ... or provide a mechanism like we do for the registry with `on_redefinition`.

Reply:
We can start with raising an error for now.
If possible we should protect default formats.